kolena/_experimental/trace/trace.py

```diff
@@ -0,0 +1,241 @@
+import atexit
+import dataclasses
+import functools
+import inspect
+import threading
+import time
+import uuid
+from collections import defaultdict
+from datetime import datetime
+from typing import Any
+from typing import Callable
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Set
+from typing import Tuple
+
+import numpy as np
+import pandas as pd
+
+import kolena
+from kolena._utils.datatypes import DataObject
+from kolena.dataset import upload_results
+from kolena.dataset.dataset import _load_dataset_metadata
+from kolena.dataset.dataset import _upload_dataset
+from kolena.errors import NotFoundError
+
+THIRTY_SECONDS = 30
+KOLENA_DEFAULT_ID = "_kolena_id"
+KOLENA_TIMESTAMP_KEY = "_kolena_timestamp"
+KOLENA_TIME_ELAPSED_KEY = "_kolena_time_elapsed"
+
+
+def _serialize_item(value: Any):
+    if isinstance(value, (bool, str, int, float)) or value is None:
+        return value
+    elif isinstance(value, np.generic):  # numpy scalars are common enough to be worth specific handling
+        for base_type, numpy_type in [(bool, np.bool_), (int, np.integer), (float, np.inexact)]:
+            if isinstance(value, numpy_type):  # cast if there is a match, otherwise fallthrough
+                return base_type(value)
+    elif isinstance(value, Dict):
+        return {key: _serialize_item(subvalue) for key, subvalue in value.items()}
+    elif isinstance(value, DataObject):
+        return value._to_dict()
+    elif dataclasses.is_dataclass(value):
+        return _serialize_item(dataclasses.asdict(value))
+    elif hasattr(value, "__dict__"):
+        return _serialize_item(value.__dict__)
+    elif isinstance(value, (List, Set, Tuple, pd.Series, np.ndarray)):
+        return [_serialize_item(item) for item in value]
+    else:
+        return str(value)
+
+
+class _Trace:
+    def __init__(
+        self,
+        func: Callable,
+        *,
+        dataset_name: Optional[str] = None,
+        model_name: Optional[str] = None,
+        model_name_field: Optional[str] = None,
+        sync_interval=THIRTY_SECONDS,
+        id_fields: Optional[List[str]] = None,
+        record_timestamp: bool = True,
+    ):
+        kolena.initialize(verbose=False)
+        self.func = func
+        self.signature = inspect.signature(func)
+        if not dataset_name:
+            dataset_name = func.__name__
+        self.dataset_name = dataset_name
+        if model_name:
+            self.model_name = model_name
+        else:
+            self.model_name = f"{self.dataset_name}_model"
+        if model_name_field:
+            if model_name_field not in self.signature.parameters:
+                raise ValueError(f"Model Name Field {model_name_field} not found in function signature")
+            self.model_name_field = model_name_field
+        try:
+            self.existing_dataset = _load_dataset_metadata(self.dataset_name)
+            if sorted(id_fields) != sorted(self.existing_dataset.id_fields):
```

Comment:
if we call a func with wrapper once without custom id_fields, we will have to specify the id_fields to be `KOLENA_DEFAULT_ID`, is this expected?

```sh
In [1]: from kolena._experimental.trace import kolena_trace

In [2]: @kolena_trace
   ...: def f():
   ...:     """this is my doc string"""
   ...:     return 42
   ...: 
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-89bdf6d431ab> in <cell line: 1>()
      1 @kolena_trace
----> 2 def f():
      3     """this is my doc string"""
      4     return 42
      5 

~/Documents/kolena-repos/kolena/kolena/_experimental/trace/trace.py in kolena_trace(func, dataset_name, model_name, model_name_field, sync_interval, id_fields, record_timestamp)
    217     """
    218     if func:
--> 219         return _Trace(
    220             func,
    221             dataset_name=dataset_name,

~/Documents/kolena-repos/kolena/kolena/_experimental/trace/trace.py in __init__(self, func, dataset_name, model_name, model_name_field, sync_interval, id_fields, record_timestamp)
     94         try:
     95             self.existing_dataset = _load_dataset_metadata(self.dataset_name)
---> 96             if sorted(id_fields) != sorted(self.existing_dataset.id_fields):
     97                 raise ValueError(f"Id Fields {id_fields} do not match existing dataset id fields")
     98             self.id_fields = self.existing_dataset.id_fields

TypeError: 'NoneType' object is not iterable

In [3]: @kolena_trace(id_fields=["_kolena_id"])
   ...: def f():
   ...:     """this is my doc string"""
   ...:     return 42
   ...: 
```

Reply:
not expected